src/compaction_filter.h

```diff
@@ -0,0 +1,142 @@
+#pragma once
+
+#include "db_impl.h"
+#include "rocksdb/compaction_filter.h"
+
+#include <utility>
+
+namespace rocksdb {
+namespace titandb {
+
+class TitanCompactionFilter final : public CompactionFilter {
+public:
+  explicit TitanCompactionFilter(const CompactionFilter *original,
+                                 std::shared_ptr<BlobStorage> blob_storage,
+                                 bool manual_release)
+      : blob_storage_(std::move(blob_storage)), original_filter_(original),
+        manual_release_(manual_release) {}
+
+  ~TitanCompactionFilter() override {
+    if (manual_release_) {
+      delete original_filter_;
+    }
+  }
+
+  const char *Name() const override {
+    return std::string("TitanCompactionFilter.")
+        .append(original_filter_->Name())
+        .c_str();
+  }
+
+  Decision FilterV2(int level, const Slice &key, ValueType value_type,
+                    const Slice &value, std::string *new_value,
+                    std::string *skip_until) const override {
+    if (value_type != kBlobIndex) {
+      return original_filter_->FilterV2(level, key, value_type, value,
+                                        new_value, skip_until);
+    }
+
+    BlobIndex blob_index;
+    Slice original_value(value.data());
+    Status s = blob_index.DecodeFrom(&original_value);
+    if (!s.ok()) {
+      // Unable to decode blob index. Keeping the value.
+      return Decision::kKeep;
+    }
+    if (BlobIndex::IsDeletionMarker(blob_index)) {
+      // TODO(yiwu): handle deletion marker at bottom level.
+      return Decision::kKeep;
+    }
+
+    BlobRecord record;
+    PinnableSlice buffer;
+
+    if (blob_storage_) {
+      ReadOptions read_options;
+      s = blob_storage_->Get(read_options, blob_index, &record, &buffer);
```

Comment:
if the status is not okay(missing blob file). We can regard it as a stale index, and return `Decision::kRemove`

Reply:
Did you mean `compaction_style`?

Reply:
change logic like this
```suggestion
      s = blob_storage_->Get(read_options, blob_index, &record, &buffer);
      if (s.IsCorruption()) {
          // meet a stale blob index, or bug. so just keep it
          return Decision::kKeep;
      } else if (s.ok()) {
          ......
      } else {
          db_->SetBGError(s);
      }
```

Reply:
I think it should be
```
if (s.ok()) {
   // return the value
} else if (s.IsNotFound()) {
   // ignore the value. The blob index has been GC-ed.
} else {
   db_->SetBGError(s);
}
```

Reply:
Unfortunately `Corruption` is an ambiguous error code where it could be a real corruption or blob file not found. We probably need to update `BlobStorage::Get` to return a flag to tell this two cases apart.